tests/ignite/engine/test_engine.py

```diff
@@ -1014,3 +1168,66 @@ def train_step(engine, batch):
     assert trainer.state.iteration == 20 * 10
     assert trainer.state.epoch == 20
     assert trainer.state.dataloader is None
+
+
+@pytest.mark.parametrize("data, epoch_length", [(None, 10), (range(10), None)])
+def test_engine_run_resume(data, epoch_length):
+    # https://github.com/pytorch/ignite/wiki/Roadmap#runresume-logic-improvements
+    engine = Engine(lambda e, b: None)
+    real_epoch_length = len(data) if data is not None else epoch_length
+
+    first_epoch_iter = [None, None]
+
+    @engine.on(Events.STARTED, first_epoch_iter)
+    def check_iter_epoch(first_epoch_iter):
+        assert engine.state.epoch == first_epoch_iter[0]
+        assert engine.state.iteration == first_epoch_iter[1]
+
+    # (re)start from 0 to 5
+    first_epoch_iter[0], first_epoch_iter[1] = 0, 0
+    # Engine run starting with max_epochs=5 => state.epoch=5
+    engine.run(data, max_epochs=5, epoch_length=epoch_length)
+    assert engine.state.epoch == 5
+    assert engine.state.iteration == 5 * real_epoch_length
+
+    # continue from 5 to 7
+    first_epoch_iter[0], first_epoch_iter[1] = 5, 5 * real_epoch_length
+    # Engine run resuming from iteration 50, epoch 5 until 7 epochs => state.epoch=7
+    engine.run(data, max_epochs=7, epoch_length=epoch_length)
+    assert engine.state.epoch == 7
+    assert engine.state.iteration == 7 * real_epoch_length
+
+    # error
+    with pytest.raises(ValueError, match="Argument max_epochs should be larger than the start epoch"):
+        engine.run(data, max_epochs=4, epoch_length=epoch_length)
+
+    # restart from 0 to 7 (As state.epoch == max_epochs(=7),
+    # this should be like that as we always do: evaluator.run(data) without any other instructions)
+    first_epoch_iter[0], first_epoch_iter[1] = 0, 0
+    # Engine run starting with max_epochs=7 => state.epoch=7
+    engine.run(data, max_epochs=7, epoch_length=epoch_length)
```

Comment:
Here `max_epoch` is equal to `state.max_epoch` but in the engine code we have `Argument max_epochs should be larger than the start epoch defined in the state`.

Reply:
So, what is the issue with that. If provided `max_epochs  == engine.state.max_epochs` then engine will restart. Similarly, to `evaluator.run(val_data)` where max_epochs=1 and after the first validation run we have `evaluator.state.max_epochs==1`.

Reply:
So in docstring we should say `Argument max_epochs should be greater than or equal to the start epoch defined in the state`.

Reply:
Yes, "Argument max_epochs should be greater or equal to the start epoch defined in the state"
if you would like to send a PR with the change, feel free.

Reply:
Couldn't we do in this same PR?

Reply:
Let's fix that in a follow-up PR